source/grid/grid_generator.cc

```diff
@@ -8403,6 +8403,209 @@ namespace GridGenerator
 
 
 
+  template <int dim, int spacedim>
+  void
+  alfeld_split_of_simplex_mesh(const Triangulation<dim, spacedim> &in_tria,
+                               Triangulation<dim, spacedim>       &out_tria)
+  {
+    Assert(dim == 2, ExcNotImplemented());
+
+    Triangulation<dim, spacedim> temp_tria;
+    if (in_tria.n_global_levels() > 1)
+      {
+        AssertThrow(!in_tria.has_hanging_nodes(), ExcNotImplemented());
+        GridGenerator::flatten_triangulation(in_tria, temp_tria);
+      }
+    const Triangulation<dim, spacedim> &ref_tria =
+      in_tria.n_global_levels() > 1 ? temp_tria : in_tria;
+
+    // Three triangles connecting to barycenter with vertex index 3:
+    static const ndarray<unsigned int, 3, 3> table_2D_cell = {
+      {{{0, 1, 3}}, {{1, 2, 3}}, {{2, 0, 3}}}};
+
+    /* Boundary-faces 2d:
+     * After converting, each of the 4 quadrilateral faces is defined by faces
+     * of 2 different triangles, i.e., lines. Note that lines are defined by 2
+     * vertices.
+     */
+    static const ndarray<unsigned int, 4, 2, 2>
+      vertex_ids_for_boundary_faces_2d = {
+        {{{{{0, 1}}}}, {{{{1, 2}}}}, {{{{2, 0}}}}}};
+
+
+    std::vector<Point<spacedim>> vertices;
+    std::vector<CellData<dim>>   cells;
+    SubCellData                  subcell_data;
+
+    // for each vertex we store the assigned index so that we only
+    // assign them a value once
+    std::vector<unsigned int> old_to_new_vertex_indices(
+      ref_tria.n_vertices(), numbers::invalid_unsigned_int);
+
+    // We first have to create all of the new vertices. To do this, we loop over
+    // all cells and on each cell
+    // (i) copy the existing vertex locations (and record their new indices in
+    // the 'old_to_new_vertex_indices' vector),
+    // (ii) create new barycenter vertex location
+    for (const auto &cell : ref_tria.cell_iterators())
+      {
+        // temporary array storing the global indices of each cell entity in the
+        // sequence: vertices, edges/faces, cell
+        std::array<unsigned int, dim == 2 ? 4 : 5> local_vertex_indices;
+
+        // (i) copy the existing vertex locations
+        Tensor<1, spacedim> barycenter;
+        for (const auto v : cell->vertex_indices())
+          {
+            const auto v_global = cell->vertex_index(v);
+
+            if (old_to_new_vertex_indices[v_global] ==
+                numbers::invalid_unsigned_int)
+              {
+                old_to_new_vertex_indices[v_global] = vertices.size();
+                vertices.push_back(cell->vertex(v));
+              }
+
+            AssertIndexRange(v, local_vertex_indices.size());
+            local_vertex_indices[v] = old_to_new_vertex_indices[v_global];
+
+            barycenter += vertices[local_vertex_indices[v]] - Point<spacedim>();
```

Comment:
You don't have to compute the barycenter yourself. You can just call `cell->barycenter()`.